### ISSUE TITLE
Images not visible in item display & performance issues

### ISSUE THREAD

**Issue by tdodson** [projectblacklight/spotlight#1759]
I'm evaluating Blacklight/Spotlight for use for some digital collections at the Harvard Library and have it installed on my local machine, a Mac Powerbook. I find that when I manually upload image items they show up in the list of items, but only sometimes display the thumbnails (or sometimes they appear after a few seconds but with no indication from the browser or any UI element that they are loading)

When I go to review the images, the metadata displays, as does the border/pane and controls for zooming, etc., but most of the time not the image itself.

<img width="1001" alt="screen shot 2017-02-23 at 9 34 46 am" src="https://cloud.githubusercontent.com/assets/1158342/23263155/5a57a2be-f9ab-11e6-82b3-bc3f18eed767.png">.
 
In addition, the application runs very slowly, with pages taking ~10 seconds to load in some cases. Do you think this is because I'm running it on a local machine or has performance been an ongoing problem with the application even on more robust servers? Thanks.

Tom
---
Thomas Dodson
Web Developer & Designer
Harvard Library

**Comment by jkeck** [comment]
Hi Thomas,
I think what you may be seeing is poor performance from your local development environment trying to simultaneously host the rails application as well as the IIIF compatible image server we ship with Spotlight (RIIIF).  This can potentially be slow in a development environment (particularly depending on your local web application server and how many threads it can run).

Are you noticing any error in particular from either your rails logs or your javascript console in your web browser?  That could indicate long running requests or potentially a larger problem that we could help debug.

FWIW, we certainly haven’t seen > 10s load times for pages in our recent local development (on modern Macbook Pros), and we are running a production Spotlight application at exhibits.stanford.edu<http://exhibits.stanford.edu> (which is running on relatively robust servers).

- Jessie

On Feb 23, 2017, at 6:35 AM, Thomas Dodson <notifications@github.com<mailto:notifications@github.com>> wrote:


I'm evaluating Blacklight/Spotlight for use for some digital collections at the Harvard Library and have it installed on my local machine, a Mac Powerbook. I find that when I manually upload image items they show up in the list of items, but only sometimes display the thumbnails (or sometimes they appear after a few seconds but with no indication from the browser or any UI element that they are loading)

When I go to review the images, the metadata displays, as does the border/pane and controls for zooming, etc., but most of the time not the image itself.

[screen shot 2017-02-23 at 9 34 46 am]<https://cloud.githubusercontent.com/assets/1158342/23263155/5a57a2be-f9ab-11e6-82b3-bc3f18eed767.png>.

In addition, the application runs very slowly, with pages taking ~10 seconds to load in some cases. Do you think this is because I'm running it on a local machine or has performance been an ongoing problem with the application even on more robust servers? Thanks.

Tom

Thomas Dodson
Web Developer & Designer
Harvard Library

**Comment by jcoyne** [comment]
@tdodson are you running on a multithreaded app server such as puma or a single threaded server such as webrick?

**Comment by tdodson** [comment]
I'm using puma. I'll take a closer look at the logs and see if I can determine what's going on. Thanks for the fast and helpful responses to questions I've posted here.

**Comment by tdodson** [comment]
The only error message I can find that rails is giving me is this one: ActionController::RoutingError (No route matches [GET] "/assets/openseadragon/openseadragon.js.map"):

I receive a message in Javascript console of: openseadragon.self-77011f3….js?body=1:19750 Tile $.Tile failed to load: http://localhost:3000/images/4/full/10,/0/default.jpg - error: Image load exceeded timeout

**Comment by jcoyne** [comment]
You aren't really using a PowerBook (last manufactured in2006) are you?

**Comment by danschmidt5189** [comment]
I'm running into the same issue with a production installation (puma behind nginx, with nginx serving static assets), and the issue is image size — the images are _8-12mb_ and thus taking a long time to transfer.

@tdodson Try checking the network tab of your browser's developer toolbox. The JS error you pasted implies that load time is the issue, but the network tab would should it for sure.

@jkeck Assuming that's the issue, then rendering a loading/spinner icon while the image is being fetched from the server would be very nice. But would that be a Spotlight issue? Or... OpenSeadragon?

**Comment by jkeck** [comment]
@danschmidt5189 if we're talking about images in search results, browse categories, sir-trevor widgets, etc, then that would be in Spotlight.  If you're talking about the image viewer on the show/detail page, then that would be OpenSeadragon.

**Comment by danschmidt5189** [comment]
@jkeck I'm seeing that in a page containing multiple "Item Embed" widgets. [This page](http://exhibits.lib.berkeley.edu/literatura-de-cordel/feature/the-exhibit-narrative) provides an example — images take a long time to load and don't render on first-view even after loading, however they _do_ render once they're cached in your browser and you reload the page.

---

That same page also demos another bug — with multiple zoomable images on the page (multiple "Item Embeds"), only the first set of OpenSeadragon buttons (zoom, expand, ...) actually works, and it applies to _all_ images on the page.

This is because all buttons on the page share the same ID (e.g. "osd-zoom-in"). Even though the images themselves have unique IDs — Openseadragon1, Openseadragon2, ... — they each refer to the same button IDs and thus only the first set on the page actually works.

Thoughts? Separate issue?

**Comment by jkeck** [comment]
Yeah, I believe those are both, separate, OpenSeadragon issues (of which we've seen many).  We don't run into this problem too often locally because we don't really use OpenSeadragon in our implementation, which is why we have never tacked this before.  We may just need to update our version of OpenSeadragon as they may have addressed these things already.

On a side-note, you may be interested in some work that will be happening over the next few weeks at Stanford to easily allow [Mirador](projectmirador.org) to be used as the viewer in Spotlight instead of OpenSeadragon (which should hopefully also solve several of these issues).
